### PR TITLE
Fix issue with uri versioning (#1529) and Version annotation conditio…

### PR DESCRIPTION
**Pull request by juillerat**
Fix issues with uri versioning (#1529) and Version annotation conditions (#1491)

### CODE REVIEW

**Comment by juillerat** [comment]
In this proposition, the VersionListener does not use the ViewHandlerInterface anymore. Regardless of this, I didn't changed the signature of the constructor ...

File: Tests/Fixtures/Etalon/annotated_version_controller.yml [hunk 1]
```diff
@@ -15,3 +15,9 @@ v3_user:
   controller:   ::v3UserAction
   methods: [GET]
   condition: "request.attributes.get('version') in ['v1', 'v3']"
+
+get_users:
+  path:      /{version}/users.{_format}
+  controller:   ::getUsersAction
+  requirements: {version: 'v1|v3'}
```
Comment (GuilhemN) [inline Tests/Fixtures/Etalon/annotated_version_controller.yml]: `{ version: v1|v3 }`

File: Routing/Loader/Reader/RestActionReader.php [hunk 2]
```diff
@@ -317,7 +318,13 @@ public function read(RestRouteCollection $collection, \ReflectionMethod $method,
                 $defaults = array_merge($defaults, $annotation->getDefaults());
                 $host = $annotation->getHost();
                 $schemes = $annotation->getSchemes();
-                $combinedCondition = $this->combineConditions($versionCondition, $annotation->getCondition());
+
+                if ($this->hasVersionPlaceholder($path)) {
+                    $combinedCondition = $annotation->getCondition();
+                    $requirements = array_merge($versionRequirement, $requirements);
```
Comment (GuilhemN) [inline Routing/Loader/Reader/RestActionReader.php]: Isn't this already supported?

File: EventListener/VersionListener.php [hunk 3]
```diff
@@ -22,13 +21,11 @@
  */
 class VersionListener
 {
-    private $viewHandler;
     private $versionResolver;
     private $defaultVersion;
 
-    public function __construct(ViewHandlerInterface $viewHandler, VersionResolverInterface $versionResolver, $defaultVersion = null)
+    public function __construct(ViewHandlerInterface $viewHandler = null, VersionResolverInterface $versionResolver, $defaultVersion = null)
```
Comment (GuilhemN) [inline EventListener/VersionListener.php]: Can be removed, it is `@internal`.

File: EventListener/VersionExclusionListener.php [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+    public function onKernelRequest(GetResponseEvent $event)
+    {
+        $request = $event->getRequest();
+
+        if (!$request->attributes->get(FOSRestBundle::ZONE_ATTRIBUTE, true)) {
+            return;
+        }
+
+        if (!$request->attributes->has('version')) {
```
Comment (GuilhemN) [inline EventListener/VersionExclusionListener.php]: I'd be in favor of merging checks here.

**Comment by GuilhemN** [comment]
Sorry for answering so lately, I did not have the time to review your PR until now...

**Comment by xabbuh** [comment]
@juillerat Do you think you will be able to finish here? Otherwise, please feel free to say so and we will happily finish this one for you. :)

**Comment by winkelsdorf** [comment]
Looks good and fixes #1691 for me.

One thing to note: As soon as `{version}` is used as part of the URI, other resolvers are ignored, i.e. `Version` always resolves to the URI part. Tested by adding `query`, `custom_header` and `media_type`.

Consider a situation where an explicit URI points to a version, while using the URI without prefix should point to `latest`:

```yaml
api_with_prefix:
    resource:     ApiBundle\Controller\ProductsController
    type:         rest
    prefix:       /{version}
    name_prefix:  api_

api_latest:
    resource:     ApiBundle\Controller\ProductsController
    type:         rest
    defaults:
        version:  latest
    name_prefix:  api_latest_
```

Probably this should respect the order of the resolvers somehow, too?

**Comment by winkelsdorf** [comment]
As mentioned in #1691 I have further Issues with the proposed PR, but the current handling is really somehow broken, too.

## Consider the following situation
I want to provide two ways of versioning (the usual: easier client consumption vs precise best practice approach). One URI based and another one using the usual `custom_header` and `media_type` resolvers, the latter pointing to the latest version when no version is specified, i.e. no header or media type given.

Where possible, I want to make use of Annotations to achieve this, with the least code overhead.

## Setup

A local `parameter.yml` is loaded from my Bundle via the `config.yml` `imports`, versioning is enabled and has a default version.

**app/config/config.yml**

```yaml
imports:
    - { resource: parameters.yml }
    - { resource: security.yml }
    - { resource: services.yml }
    - { resource: "@ApiBundle/Resources/config/parameters.yml" }
...
fos_rest:
    versioning:
        enabled: true
        default_version: "%api.version%"
    format_listener: # determine best format for the request
        enabled: false
        # rules:
        #    - { path: '^/', priorities: ['json', 'xml'], fallback_format: json, prefer_extension: false }
    view:
        view_response_listener: force # allow returning a view instance from action controllers
        formats:
            json: true
            xml:  true
        mime_types:
            json: ['application/json', 'application/json; version=v1', 'application/json; version=v2']
            xml:  ['application/xml', 'application/xml; version=v1', 'application/xml; version=v2']
```

**src/ApiBundle/Resources/config/parameter.yml**

```yaml
parameters:
    api.versions: v1|v2 # available versions
    api.version:  v2 # current version
```

Routing is directed to the Bundle in  
**app/config/routing.yml**

```yaml
api:
    resource: "@ApiBundle/Resources/config/routing.yml"
    type:     rest
```

Now the routing inside the Bundle:  
**src/ApiBundle/Resources/config/routing.yml**

```yaml
api_with_prefix:
    resource:     "@ApiBundle/Controller/"
    type:         rest
    prefix:       /{version}          # is this mapped to the Version Annotation?
    requirements:
        version:  "%api.versions%"    # allowed path versions from local parameters.yml
    name_prefix:  api_

api_latest_overridable:
    resource:     "@ApiBundle/Controller/"
    type:         rest
    name_prefix:  api_latest_
```

New to Symfony & FOSRestBundle, please correct me if something is wrong with the above configuration or my expectations.

## Directory Structure

```console
src\ApiBundle\Controller\v1\
src\ApiBundle\Controller\v1\ProductController.php

src\ApiBundle\Controller\v2\
src\ApiBundle\Controller\v2\ProductController.php
```

## Controllers

**src\ApiBundle\Controller\v1\ProductController.php**

```php
<?php

namespace ApiBundle\Controller\v1;

use FOS\RestBundle\Controller\Annotations\NamePrefix;
use FOS\RestBundle\Controller\Annotations\RouteResource;
use FOS\RestBundle\Controller\Annotations\Version;
use FOS\RestBundle\Controller\FOSRestController;
use FOS\RestBundle\Routing\ClassResourceInterface;
use Symfony\Component\HttpFoundation\Request;

/**
 * Class ProductController
 * @RouteResource("Product")
 * @Version("v1")
 * @NamePrefix("v1_")
 */
class ProductController extends FOSRestController implements ClassResourceInterface
{
    /**
     * c stands for Collection
     * /products
     * @param Request $request
     * @return array
     */
    public function cgetAction(Request $request)
    {
        $requestVersion = $request->get("version");
        $defaultVersion = $this->getParameter("api.version");
        $content = array(
            'Controller' => '1',
            'Requested' => $requestVersion,
            # 'Default' => $defaultVersion
        );

        return $content;
    }
    
    /**
     * /products/{$id}
     * @param $id
     */
    public function getAction($id)
    {
    }
}
```

For this test Version 2 Controller is basically the same, except for all `1` changed to `2`. 

`@RouteResource` and `ClassResourceInterface` seem to generate the same resources. As the 2nd part of the PR is about `ClassResourceInterface` Versioning it is kept. Behavior is the same with and without.

**src\ApiBundle\Controller\v2\ProductController.php**

```php
<?php

namespace ApiBundle\Controller\v2

use FOS\RestBundle\Controller\Annotations\NamePrefix;
use FOS\RestBundle\Controller\Annotations\RouteResource;
use FOS\RestBundle\Controller\Annotations\Version;
use FOS\RestBundle\Controller\FOSRestController;
use FOS\RestBundle\Routing\ClassResourceInterface;
use Symfony\Component\HttpFoundation\Request;

/**
 * Class ProductController
 * @RouteResource("Product")
 * @Version("v2")
 * @NamePrefix("v2_")
 */
class ProductController extends FOSRestController implements ClassResourceInterface
{
    /**
     * c stands for Collection
     * /products
     * @param Request $request
     * @return array
     */
    public function cgetAction(Request $request)
    {
        $requestVersion = $request->get("version");
        $defaultVersion = $this->getParameter("api.version");
        $content = array(
            'Controller' => '2',
            'Requested' => $requestVersion,
            # 'Default' => $defaultVersion
        );

        return $content;
    }
    
    /**
     * /products/{$id}
     * @param $id
     */
    public function getAction($id)
    {
    }
}
```

Basically both test controllers should return the actively used version number, the version requested and the default API version.

## Expectations and actual results

As described in the scenario, I would expect the following to work but in some cases the wrong controller is matched.

The Header Variants for unprefixed URIs:

| URI | Accept Header | Expected | Actual |
|---|---|---|---|
| /products | w/o | (Default API): Controller 2, Requested v2 | correct |
| /products | X-Accept-Version: v1 | Controller 1, Requested v1 | correct |
| /products | Accept: application/json; version=v1 | Controller 1, Requested v1 | **Wrong** returns: Controller 2, Requested v2 |

The prefixed URI variants:

| URI | Accept Header | Expected | Actual |
|---|---|---|---|
| /v1/products | w/o | Controller 1, Requested v1 | **Wrong** returns: Controller 2, Requested v1 |
| /v2/products | w/o | Controller 2, Requested v2 | correct |

Addendum: Furthermore the prefixed URI variants should *still* respect the serialization format sent via the `Accept-Header`. This is not tested here. 

## Observations

- The described wrong **behavior is unchanged** for the master branch and the PR fork.
- **Additional Error**: Removing the `default_version` from `config.yml` leads to 404 `NotFoundHttpException` for all above routes. Even when `defaults:` `version: "%api.version%"` is set for each route in local `routing.yml`.
- It is **not possible to debug** routes via e.g. `php bin/console router:match /v2/products` as soon as `@Version` annotation is used. Console returns with `[RuntimeException] Unable to get a property on a non-object.` - all this while the route works (but sometimes returns the wrong controller as stated in the above test matrix).

Routes are generated correctly all the time:

```console
 ~/Projects/GitHub/Test-Rest-API   master ●✚  php bin/console debug:router
 ---------------------------- -------- -------- ------ ----------------------------------- 
  Name                         Method   Scheme   Host   Path                               
 ---------------------------- -------- -------- ------ ----------------------------------- 
  _wdt                         ANY      ANY      ANY    /_wdt/{token}                      
  _profiler_home               ANY      ANY      ANY    /_profiler/                        
  _profiler_search             ANY      ANY      ANY    /_profiler/search                  
  _profiler_search_bar         ANY      ANY      ANY    /_profiler/search_bar              
  _profiler_info               ANY      ANY      ANY    /_profiler/info/{about}            
  _profiler_phpinfo            ANY      ANY      ANY    /_profiler/phpinfo                 
  _profiler_search_results     ANY      ANY      ANY    /_profiler/{token}/search/results  
  _profiler_open_file          ANY      ANY      ANY    /_profiler/open                    
  _profiler                    ANY      ANY      ANY    /_profiler/{token}                 
  _profiler_router             ANY      ANY      ANY    /_profiler/{token}/router          
  _profiler_exception          ANY      ANY      ANY    /_profiler/{token}/exception       
  _profiler_exception_css      ANY      ANY      ANY    /_profiler/{token}/exception.css   
  _twig_error_test             ANY      ANY      ANY    /_error/{code}.{_format}           
  index                        ANY      ANY      ANY    /                                  
  api_v1_get_products          GET      ANY      ANY    /{version}/products                
  api_v1_get_product           GET      ANY      ANY    /{version}/products/{id}           
  api_v2_get_products          GET      ANY      ANY    /{version}/products                
  api_v2_get_product           GET      ANY      ANY    /{version}/products/{id}           
  api_latest_v1_get_products   GET      ANY      ANY    /products                          
  api_latest_v1_get_product    GET      ANY      ANY    /products/{id}                     
  api_latest_v2_get_products   GET      ANY      ANY    /products                          
  api_latest_v2_get_product    GET      ANY      ANY    /products/{id}                     
  nelmio_api_doc_index         GET      ANY      ANY    /doc/{view}                        
 ---------------------------- -------- -------- ------ ----------------------------------- 
```

## Test Setup

All tests were run with the current tagged release and this PR using an adjusted `composer.json` in the latter case:

```json
    "repositories": [
        {
            "type": "vcs",
            "url": "https://github.com/juillerat/FOSRestBundle.git"
        }
    ],
    "require": {
        "php": ">=5.5.9",
        "symfony/symfony": "3.2.*",
        "doctrine/orm": "^2.5",
        "doctrine/doctrine-bundle": "^1.6",
        "doctrine/doctrine-cache-bundle": "^1.2",
        "symfony/swiftmailer-bundle": "^2.3.10",
        "symfony/monolog-bundle": "^3.0.2",
        "symfony/polyfill-apcu": "^1.0",
        "sensio/distribution-bundle": "^5.0",
        "sensio/framework-extra-bundle": "~3.0",
        "incenteev/composer-parameter-handler": "^2.0",
        "twig/twig": "^1.0||^2.0",
        "friendsofsymfony/rest-bundle": "dev-master",
        "nelmio/api-doc-bundle": "^2.13"
    },
```

I verified that the in this PR referenced commit #2726af9 from the repo of @juillerat is used when running `composer update`.

**Comment by mitrae** [comment]
any news?

File: EventListener/VersionListener.php [hunk 5]
```diff
@@ -22,13 +21,11 @@
...
     {
-        $this->viewHandler = $viewHandler;
```
Comment (mitrae) [inline EventListener/VersionListener.php]: @juillerat why? VersionExclusionListener sets version into viewHandler. Wouldn't be logical to take it here?

**Comment by mitrae** [comment]
even with this PR I still get exception "Unable to get property on a non-object"
but looks like it is bug in symfony itself https://github.com/symfony/symfony/pull/22636

**Comment by masterkaos** [comment]
Any news on this PR? I was bashing my head into the wall for the majority of the day trying to figure out why this wasnt working, then came across this. Sad to see this  issue open for so long when it seems like something that would be pretty common to run into...

**Comment by mmoll** [comment]
@juillerat this would need a rebase...

**Comment by bilel-noomene** [comment]
+1

**Comment by luiztiago** [comment]
👍

**Comment by GuilhemN** [comment]
Closing in favor of https://github.com/FriendsOfSymfony/FOSRestBundle/pull/1937, thank you!

**Pull request closed without merging**